### PR TITLE
Improved the code in IF and TRAK to support passing in dict

### PR DESCRIPTION
**Pull request by Suliang-Jin**
## Description

I have changed some code in IF, TRAK and TracIn so they now support passing in dict. If the user still passes in tuple or list, the behavior doesn't change.
I have also added some experiments on supporting IF and Huggingface transformers.

### CODE REVIEW

**Review by jiaqima — COMMENTED** [review]
It seems that this PR contained changes from another PR #217. If the proposed changes in this PR depend on the other PR, please make this PR after merging the other one.

File: dattri/algorithm/base.py [hunk 1]
```diff
@@ -345,9 +345,17 @@ def attribute(
...
-                train_batch_data = tuple(
-                    data.to(self.device).unsqueeze(0) for data in train_batch_data_
-                )
+                if isinstance(train_batch_data_, (tuple, list)):
+                    train_batch_data = tuple(
+                        data.to(self.device).unsqueeze(0) for data in train_batch_data_
+                    )
+                elif isinstance(train_batch_data_, dict):
+                    train_batch_data = {
+                        k: v.unsqueeze(0) for k, v in train_batch_data_.items()
```
Comment (TheaperDeng) [inline dattri/algorithm/base.py]: We also assume the value in dictionary to be tensor right?

File: docs/guide/README.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
+# User Guide
```
Comment (TheaperDeng) [inline docs/guide/README.md]: The documentation is clear, a high-level summary table at the top would be beneficial. It should list the supported data types and callable types for all methods in the https://github.com/TRAIS-Lab/dattri?tab=readme-ov-file#supported-algorithms.

File: docs/guide/README.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+    outputs = torch.func.functional_call(
+        model,
+        params,
+        input_ids,
+        kwargs={"attention_mask": attention_mask, "labels": labels},
+    )
+    logp = -outputs.loss
+    return logp - torch.log(1 - torch.exp(logp))
+```
```
Comment (TheaperDeng) [inline docs/guide/README.md]: Slightly different requirements should be applied to TRAK (Multi-class Margin) and TracIN (training loss and any target function).

File: experiments/gpt2_wikitext/score_IF_dict.py [hunk 4]
```diff
@@ -0,0 +1,686 @@
+#!/usr/bin/env python
```
Comment (TheaperDeng) [inline experiments/gpt2_wikitext/score_IF_dict.py]: We only need one script for IF

File: experiments/gpt2_wikitext/score_TRAK_dict.py [hunk 5]
```diff
@@ -0,0 +1,742 @@
+#!/usr/bin/env python
```
Comment (TheaperDeng) [inline experiments/gpt2_wikitext/score_TRAK_dict.py]: We only need one script for TRAK.

File: docs/guide/README.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+|                                                   |             [DataInf](https://arxiv.org/abs/2310.00902)              | ✔️ | ✔️ | ❌ | [Code example](../../examples/brittleness/mnist_lr_brittleness.py) |
+|                                                   |              [EK-FAC](https://arxiv.org/abs/2308.03296)              | ✔️ | ✔️ | ❌ | [Code example](../../examples/brittleness/mnist_lr_brittleness.py) |
+|                                                   |             [RelatIF](https://arxiv.org/pdf/2003.11630)              | ✔️ | ✔️ | ❌ | [Code example](../../examples/brittleness/mnist_lr_brittleness.py) |
+|                                                   |              [LoGra](https://arxiv.org/pdf/2405.13954)               | ✔️ | ✔️ | ❌ | [Code example](../../examples/brittleness/mnist_lr_brittleness.py) |
+|                                                   |              [GraSS](https://arxiv.org/pdf/2505.18976)               | ✔️ | ✔️ | ❌ | [Code example](../../examples/brittleness/mnist_lr_brittleness.py) |
```
Comment (TheaperDeng) [inline docs/guide/README.md]: I think for GraSS, LoGra, RelateIF, EK-FAC, we don't have their examples in ../../examples/brittleness/mnist_lr_brittleness.py

File: experiments/gpt2_wikitext/score_TRAK.py [hunk 7]
```diff
@@ -312,6 +312,13 @@ def parse_args():
             "Use 'TRAK-k' to load k checkpoints and run TRAK."
         ),
     )
+
+    parser.add_argument(
+        "--data_structure",
+        type=str,
+        default="tuple",
+        choices=["tuple", "list", "dict"]
+    )
```
Comment (TheaperDeng) [inline experiments/gpt2_wikitext/score_TRAK.py]: We don't need to show-off what we have supported in the examples. Just choose the most convenient way and demonstrate it in the script.

File: experiments/gpt2_wikitext/score_TRAK.py [hunk 8]
```diff
@@ -312,6 +312,13 @@ def parse_args():
             "Use 'TRAK-k' to load k checkpoints and run TRAK."
         ),
     )
+
+    parser.add_argument(
+        "--data_structure",
+        type=str,
+        default="tuple",
+        help="What data structure to pass the training/test data for data attribution."
+    )
```
Comment (sx-liu) [inline experiments/gpt2_wikitext/score_TRAK.py]: We could simply remove this argument, and assume the input structure is `dict`, as it is more natural for huggingface datasets.

File: experiments/gpt2_wikitext/score_TRAK.py [hunk 9]
```diff
@@ -591,94 +598,163 @@ def group_texts(examples):
...
+            input_ids = input_ids.cuda()
+            attention_mask = attention_mask.cuda()
+            labels = labels.cuda()
+            outputs = torch.func.functional_call(
+                model,
+                params,
+                input_ids,
+                kwargs={"attention_mask": attention_mask, "labels": labels},
+            )
+            return outputs.loss
```
Comment (sx-liu) [inline experiments/gpt2_wikitext/score_TRAK.py]: Same here

File: test/dattri/algorithm/test_influence_function_dict.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        train_loader = DataLoader(train_data, batch_size=4)
+        test_loader = DataLoader(test_data, batch_size=2)
+
+        def f(params, batch):
+            image = batch["image"]
+            label = batch["label"]
+            loss = nn.CrossEntropyLoss()
+            yhat = torch.func.functional_call(model, params, image)
+            return loss(yhat, label.long())
+
```
Comment (sx-liu) [inline test/dattri/algorithm/test_influence_function_dict.py]: Can we also include the test case for IF Explicit?
